### PR TITLE
Uncomplete blocks dependent on an answer that changes

### PR DESCRIPTION
**Pull request by dwyeradam**
### What is the context of this PR?
If I answer the 1st question as 100 and then the 2nd page questions with appropriate answers, if i go back and edit the 1st question to 1 and jump back to the summary page.
I am able to submit the answers in an invalid state.
I think if we modify an answer that is depended on by validation then we should invalidate the answers for the dependent block.

### CODE REVIEW

**Review by yoganandkunche — APPROVED** [review]

File: app/questionnaire/questionnaire_schema.py [hunk 1]
```diff
@@ -167,6 +175,42 @@ def _get_aliases(self):
 
         return aliases
 
+    def get_answer_dependencies(self):
```
Comment (jonnyshaw89) [inline app/questionnaire/questionnaire_schema.py]: theses should be private `_get_answer_dependencies(self):`

File: app/questionnaire/questionnaire_schema.py [hunk 2]
```diff
@@ -167,6 +175,42 @@ def _get_aliases(self):
...
+        for question in self._questions_by_id.values():
+            for calculation in question.get('calculations', []):
+                answer_id = calculation.get('answer_id')
+                if answer_id:
+                    for dependency_id in calculation['answers_to_calculate']:
+                        dependencies[answer_id] = dependencies.get(answer_id, []) + [dependency_id]
+
+        return dependencies
+
+    def get_answer_block_group_ids(self):
```
Comment (jonnyshaw89) [inline app/questionnaire/questionnaire_schema.py]: theses should be private `_get_answer_block_group_ids(self):`

File: app/views/questionnaire.py [hunk 3]
```diff
@@ -566,12 +566,24 @@ def update_questionnaire_store_with_form_data(questionnaire_store, location, ans
                 _remove_answer_from_questionnaire_store(answer_id, questionnaire_store)
 
             if answer:
+                latest_answer_store_hash = questionnaire_store.answer_store.get_hash()
                 questionnaire_store.answer_store.add_or_update(answer)
+                _remove_dependent_answers_from_completed_blocks(answer_id, questionnaire_store, latest_answer_store_hash)
```
Comment (jonnyshaw89) [inline app/views/questionnaire.py]: i think that `if latest_answer_store_hash != questionnaire_store.answer_store.get_hash():` should wrap this function call.
otherwise it hides that is it not always run

**Review by jonnyshaw89 — DISMISSED** [review]

File: app/views/questionnaire.py (hunk 3 above)
Comment (dwyeradam) [reply app/views/questionnaire.py]: I did that originally but it would fail lint as too complex. So I cheated.

Comment (jonnyshaw89) [reply app/views/questionnaire.py]: tut tut

Comment (dwyeradam) [reply app/views/questionnaire.py]: when are you going?

**Comment by dwyeradam** [comment]
@nealedj Could you review this please and if you're happy I've addressed his comments dismiss Jon's review.

File: app/questionnaire/questionnaire_schema.py [hunk 4]
```diff
@@ -167,6 +175,42 @@ def _get_aliases(self):
 
         return aliases
 
+    def _get_answer_dependencies(self):
+        dependencies = {}
+        # Answer level dependencies
+        for answer in self._answers_by_id.values():
```
Comment (nealedj) [inline app/questionnaire/questionnaire_schema.py]: Might be cleaner to make use of the `answers` property.

```
for answer in self.answers:
    ...
```

File: app/questionnaire/questionnaire_schema.py [hunk 5]
```diff
@@ -167,6 +175,42 @@ def _get_aliases(self):
...
+            dependency_id = answer['id']
+            if 'min_value' in answer and 'answer_id' in answer['min_value']:
+                answer_id = answer['min_value']['answer_id']
+                dependencies[answer_id] = dependencies.get(answer_id, []) + [dependency_id]
+            if 'max_value' in answer and 'answer_id' in answer['max_value']:
+                answer_id = answer['max_value']['answer_id']
+                dependencies[answer_id] = dependencies.get(answer_id, []) + [dependency_id]
+
+        # Question level dependencies
+        for question in self._questions_by_id.values():
```
Comment (nealedj) [inline app/questionnaire/questionnaire_schema.py]: Same here.

```
for question in self.questions:
    ...
```

File: app/questionnaire/questionnaire_schema.py [hunk 6]
```diff
@@ -167,6 +175,42 @@ def _get_aliases(self):
...
         return aliases
 
+    def _get_answer_dependencies(self):
+        dependencies = {}
+        # Answer level dependencies
+        for answer in self._answers_by_id.values():
+            dependency_id = answer['id']
+            if 'min_value' in answer and 'answer_id' in answer['min_value']:
+                answer_id = answer['min_value']['answer_id']
+                dependencies[answer_id] = dependencies.get(answer_id, []) + [dependency_id]
```
Comment (nealedj) [inline app/questionnaire/questionnaire_schema.py]: Consider using a `defaultdict(list)` for this instead.

```
dependencies = defaultdict(list)
if 'min_value' in answer and 'answer_id' in answer['min_value']:
    answer_id = answer['min_value']['answer_id']
    dependencies[answer_id].append(dependency_id)
```

File: app/views/questionnaire.py [hunk 7]
```diff
@@ -542,36 +542,48 @@ def update_questionnaire_store_with_form_data(questionnaire_store, location, ans
...
+    if not answer_value_empty(answer_value):
+        # Dates are comprised of 3 string values
+        formatted_answer_value = _format_answer_value(answer_value)
+        if formatted_answer_value:
+            return formatted_answer_value
+
+    return None
+
+
+def _remove_dependent_answers_from_completed_blocks(answer_id, questionnaire_store):
```
Comment (nealedj) [inline app/views/questionnaire.py:577]: This removes the blocks containing dependent answers but doesn't remove the answers. Is that expected?

Could we add a docstring to this function to make it clear what the intention is?

File: app/questionnaire/questionnaire_schema.py (hunk 5 above)
Comment (nealedj) [inline app/questionnaire/questionnaire_schema.py]: I don't think this solution currently identifies dependent questions which are dependent due to piped values. Is that in scope for this PR?

**Review by nealedj — CHANGES_REQUESTED** [review]

File: app/views/questionnaire.py (hunk 7 above)
Comment (dwyeradam) [reply app/views/questionnaire.py:577]: We wouldn't want to remove their answers as they could still be correct. This just means that block has to be revisited and therefore validated.

File: app/questionnaire/questionnaire_schema.py (hunk 5 above)
Comment (dwyeradam) [reply app/questionnaire/questionnaire_schema.py]: Jon and I did discuss this quite a bit but maybe we need to discuss as a whole. Piping alone would not result in any new validation being done on dependent answers and so forcing users to revisit the block would have far less benefit. We also felt that if an answer did affect another then that answer should be in a max/min type check.

File: app/questionnaire/questionnaire_schema.py (hunk 6 above)
Comment (dwyeradam) [reply app/questionnaire/questionnaire_schema.py]: Great thanks. I was trying to find an easy way to initialise the key on first use and this is a far better way.

File: app/views/questionnaire.py (hunk 7 above)
Comment (dwyeradam) [reply app/views/questionnaire.py:577]: I'll add the docstring though.

**Comment by dwyeradam** [comment]
@nealedj Ready for re-review.

**Review by nealedj — APPROVED** [review]
Looks good!

**Comment by nealedj** [comment]
@dwyeradam LGTM - ready to squash
